### PR TITLE
Streamline decoding AndroidManifest.xml

### PR DESCRIPTION
**Pull request by sv99**
Streamline decoding AndroidManifest.xml.

Decoding AndroidManifest.xml without using ResStreamDecoderContainer.

Both methods: decodeManifest decodeManifestWithResources in fact initializes decoder in the same way.

### CODE REVIEW

**Comment by iBotPeaches** [comment]
Failure looks odd. Looks related to some changes I did yesterday here: https://github.com/iBotPeaches/Apktool/commit/3fdc06a46f56d674864f6e425b778f5dfb86f1f7, but they passed?

**Comment by iBotPeaches** [comment]
Sorry appears all builds, even old ones are busted at the moment. I've opened https://github.com/iBotPeaches/Apktool/issues/3174 and will have to fix that before any new PRs.

**Comment by sv99** [comment]
> Sorry appears all builds, even old ones are busted at the moment. I've opened #3174 and will have to fix that before any new PRs.

No problem

File: brut.apktool/apktool-lib/src/main/java/brut/androlib/res/ResourcesDecoder.java [hunk 1]
```diff
@@ -104,16 +105,24 @@ public void decodeManifest(File outDir) throws AndrolibException {
         if (hasManifest()) {
             if (mConfig.decodeResources == Config.DECODE_RESOURCES_FULL ||
                 mConfig.forceDecodeManifest == Config.FORCE_DECODE_MANIFEST_FULL) {
+                ResTable resTable = getResTable();
```
Comment (iBotPeaches) [inline brut.apktool/apktool-lib/src/main/java/brut/androlib/res/ResourcesDecoder.java:108]: Is there any reason to making an intermediate variable for the res table? Its already on `mResTable`?

**Review by iBotPeaches — CHANGES_REQUESTED** [review]
> Both methods: decodeManifest decodeManifestWithResources in fact initializes decoder in the same way.

I don't think this is true. `decodeManifest` calls `getManifestFileDecoder(false)` and `decodeManifestWithResources` calls `getManifestFileDecoder(true)`

Gotta find the ticket that led to the introduction of this, so people could force decode the manifest without parsing the resource table.

**Comment by sv99** [comment]
At first glance its true but in reality initialization the same.

```java
decodeManifest(...)

    // Duo<ResFileDecoder, AXmlResourceParser> duo = getManifestFileDecoder(false);
    ResStreamDecoderContainer decoders = new ResStreamDecoderContainer();
    AXmlResourceParser axmlParser = new AndroidManifestResourceParser();
    decoders.setDecoder("xml", new XmlPullStreamDecoder(axmlParser, getResXmlSerializer()));
    Duo<ResFileDecoder, AXmlResourceParser> duo = new Duo<>(new ResFileDecoder(decoders), axmlParser);

    ResFileDecoder fileDecoder = duo.m1;
    // Set ResAttrDecoder
    duo.m2.setAttrDecoder(new ResAttrDecoder());
    ResAttrDecoder attrDecoder = duo.m2.getAttrDecoder();
    attrDecoder.setResTable(resTable);

    ...
```
`// Set ResAttrDecoder ` - in the function `decodeManifest` 

```java
decodeManifestWithResources(...)

    // Duo<ResFileDecoder, AXmlResourceParser> duo = getManifestFileDecoder(true);
    ResStreamDecoderContainer decoders = new ResStreamDecoderContainer();
    AXmlResourceParser axmlParser = new AndroidManifestResourceParser();
    // Set ResAttrDecoder
    axmlParser.setAttrDecoder(new ResAttrDecoder());
    decoders.setDecoder("xml", new XmlPullStreamDecoder(axmlParser, getResXmlSerializer()));
    Duo<ResFileDecoder, AXmlResourceParser> duo =  new Duo<>(new ResFileDecoder(decoders), axmlParser);

    ResFileDecoder fileDecoder = duo.m1;
    ResAttrDecoder attrDecoder = duo.m2.getAttrDecoder();
    attrDecoder.setResTable(resTable);

    ...
```

`// Set ResAttrDecoder ` - in the function `getManifestFileDecoder(true)` 

All tests worked.

If resources dont parsed we have empty resource table initialized in the function `getResourceTable()`.

**Comment by sv99** [comment]
PR rebased

File: brut.apktool/apktool-lib/src/main/java/brut/androlib/res/ResourcesDecoder.java (hunk 1 above)
Comment (sv99) [reply brut.apktool/apktool-lib/src/main/java/brut/androlib/res/ResourcesDecoder.java:108]: resTable used in the 2 place:

1. in the `decodeManifest(resTable, ...`
2. in the `adjustPackageManifest(resTable, ...`

**Comment by iBotPeaches** [comment]
> All tests worked.

I've been learning the past few days that the test suite wasn't as robust as I hoped :)

Thanks for the update, I'll take another look.

File: brut.apktool/apktool-lib/src/main/java/brut/androlib/res/ResourcesDecoder.java (hunk 1 above)
Comment (iBotPeaches) [reply brut.apktool/apktool-lib/src/main/java/brut/androlib/res/ResourcesDecoder.java:108]: Sorry what I mean:

 * You have `getResTable` which gives you `mResTable` on the class itself (`ResourcesDecoder`) while loading packages if not loaded yet.
 * You set the output of that to an intermediate value `resTable` and pass onward to functions.
 
 So we have a class property `mResTable` and an intermediate `resTable` and I'm confused of the introduction of the intermediate one.

**Comment by sv99** [comment]
Rebased and removed local variable.

**Comment by iBotPeaches** [comment]
Thanks for update. I understand now that the main difference is having a blank/empty resource table if we force it. Just going to find an old ticket that added this feature and hope the sample app still exists. Want to be sure that isn't regressed.

**Comment by sv99** [comment]
This is independent from decoding process.

I think need first decode resources - they are not dependent from manifest.
And only then decode manifest. Now resources implicitly decoded in the `getResTable() `.

**Review by iBotPeaches — APPROVED** [review]
Looks good, but holding on merge till after I cut 2.8.1. Which should be Sat/Sun.

**Review by iBotPeaches — CHANGES_REQUESTED** [review]
I ran some tests today and I noticed a regression that we always print

```
➜  Optimized apktool d -r --force-manifest app-debug.apk -f
I: Using Apktool 2.8.1 on app-debug.apk
I: Loading resource table...
I: Decoding AndroidManifest.xml with only framework resources...
```

```
➜  Optimized apktool d app-debug.apk -f    
I: Using Apktool 2.8.1 on app-debug.apk
I: Loading resource table...
I: Decoding AndroidManifest.xml with only framework resources...
```

```
➜  Optimized apktool d -r app-debug.apk -f
I: Using Apktool 2.8.1 on app-debug.apk
I: Copying raw manifest...
I: Copying raw resources...
```

As you can see it doesn't matter what mode we do. It still prints "decoding AndroidManifest.xml with only framework resources..."

**Comment by sv99** [comment]
Different logger message added

**Comment by sv99** [comment]
rebased

**Review by iBotPeaches — APPROVED** [review]
Going to fix a typo post merge, but otherwise looks good now.

**Pull request merged**